app/helpers/pagelets_helper.rb

```diff
@@ -67,6 +67,83 @@ def th_content(col)
     sort col[:key], as: col[:label], default: col[:default_sort] || 'ASC'
   end
 
+  def create_table_preferences
+    table_preference = TablePreference.new(:user => User.current, :name => controller_name, :columns => @selected_columns)
+    table_preference.save
+    table_preference
+  end
+
+  def update_table_preferences(table_preference)
+    table_preference.update(:columns => @selected_columns)
+    table_preference.save
+  end
+
+  def defined_columns
+    Pagelets::Manager.pagelets_at('hosts/_list', :hosts_table_column_header).map do |pt|
+      next unless pt.opts[:key]
+      {
+        key: pt.opts[:key],
+        label: pt.opts[:label],
+        profiles: pt.profiles.map { |pr| pr.label },
+        checked: @selected_columns&.include?(pt.opts[:key].to_s),
+      }
+    end.compact
+  end
+
+  def defined_categories
+    defined_columns.map { |col| col[:profiles] }.flatten.uniq
+  end
+
+  def checked_columns(category)
+    all_checked = false
+    category.each do |column|
+      if column[:checkProps][:checked]
+        all_checked = true
+      elsif all_checked
+        all_checked = nil
+        break
+      end
+    end
+    all_checked
+  end
+
+  def columns_view
+    table_preference = User.current.table_preferences.find_by(name: controller_name)
+    table_preference ||= create_table_preferences
+    update_table_preferences(table_preference) unless table_preference.columns
+    columns = []
+    defined_categories.each do |category|
+      category_key = category.downcase
+      columns << {
+        name: category,
+        key: category_key,
+        defaultExpanded: category_key == 'general',
+        checkProps: {},
+        children: [],
+      }
+      category = columns.find { |c| c[:key] == category_key }
+      defined_columns.each do |column|
+        category[:children] << {
+          name: column[:label],
+          key: column[:key],
+          checkProps: { checked: column[:checked] },
+        }
+      end
+      category[:checkProps][:checked] = checked_columns(category[:children])
+    end
+    columns
+  end
+
+  def mount_column_selector
+    url = "#{api_users_path}/#{User.current[:id]}/table_preferences/#{controller_name}"
+    react_component("ColumnSelector", data:
+      {
+        url: url,
+        columns: columns_view,
+      }
```

Comment:
```suggestion
        columns: columns_view,
        hasPreference: @selected_columns.any?
      }
```

This can be used in UI selector to decide whether update or create a table preference.

Reply:
I think `!@selected_columns.nil?` would be better way to do this, because `@selected_columns.any?` would give `false` value for selected column